File: saraandeva/.claude/skills/saraandeva-episode/scripts/createElementViaApi.mjs
#!/usr/bin/env node
/**
 * Create a Kling element via the official AK/SK API and append its
 * element_id to content/elements_registry.json. Replaces /tmp/kling_create_element.py.
 *
 * Usage:
 *   node createElementViaApi.mjs \
 *     --name Mama \
 *     --description "Adult woman, blonde hair, family-show character" \
 *     --frontal https://storage.googleapis.com/saraandeva-kling-elements/characters/mama_front.png \
 *     --refer https://.../mama_3q.png \
 *     --refer https://.../mama_profile.png \
 *     [--tag o_102]            # default Character
 *     [--external-id mama-1]
 *
 * Tag IDs: o_101 Hottest, o_102 Character, o_103 Animal, o_104 Item,
 *          o_105 Costume, o_106 Scene, o_107 Effect, o_108 Others.
 *
 * Note: Suno requires 1-3 refer images (NOT 0). When you have only one image,
 * pass it as both --frontal and --refer (same URL twice).
 */
import fs from "node:fs";
import path from "node:path";
import { createHmac } from "node:crypto";

const PROJECT_ROOT = path.resolve(path.dirname(new URL(import.meta.url).pathname), "../../../..");
const ENV_FILE = "/Volumes/Samsung500/goreadling-production/.env.local";
const REGISTRY_FILE = path.join(PROJECT_ROOT, "content", "elements_registry.json");
const BASE = "https://api-singapore.klingai.com";
const GENRE_CAP_NAME = 20, CAP_DESC = 100;

for (const line of fs.readFileSync(ENV_FILE, "utf8").split("\n")) {
  const m = line.match(/^([A-Z_][A-Z_0-9]*)=["']?(.+?)["']?$/);
  if (m && !process.env[m[1]]) process.env[m[1]] = m[2];
}
const AK = process.env.KLING_ACCESS_KEY, SK = process.env.KLING_SECRET_KEY;
if (!AK || !SK) { console.error("missing KLING keys in .env.local"); process.exit(1); }

const argv = process.argv.slice(2);
const argFlag = (n) => { const i = argv.indexOf(`--${n}`); return i >= 0 ? argv[i + 1] : null; };
const argMulti = (n) => argv.reduce((acc, a, i) => (a === `--${n}` ? [...acc, argv[i + 1]] : acc), []);
const name = argFlag("name"), description = argFlag("description"), frontal = argFlag("frontal");
const refers = argMulti("refer");
const tag = argFlag("tag") || "o_102";
const externalId = argFlag("external-id") || `${(name || "").toLowerCase().replace(/\s+/g, "_")}-1`;
if (!name || !description || !frontal) {
  console.error("required: --name, --description, --frontal. Optional: --refer (1-3 times), --tag, --external-id");
  process.exit(1);
}
if (name.length > GENRE_CAP_NAME) { console.error(`name > ${GENRE_CAP_NAME} chars`); process.exit(1); }
if (description.length > CAP_DESC) { console.error(`description > ${CAP_DESC} chars`); process.exit(1); }

// Kling rejects refer_images with 0 entries — duplicate frontal if no refers given
const referList = refers.length ? refers : [frontal];
if (referList.length > 3) { console.error("max 3 refers"); process.exit(1); }

const b64url = (b) => Buffer.from(b).toString("base64").replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
function token() {
  const now = Math.floor(Date.now() / 1000);
  const h = b64url(JSON.stringify({ alg: "HS256", typ: "JWT" }));
  const p = b64url(JSON.stringify({ iss: AK, exp: now + 1800, nbf: now - 5 }));
  const s = b64url(createHmac("sha256", SK).update(`${h}.${p}`).digest());
  return `${h}.${p}.${s}`;
}
async function http(method, p, body = null) {
  const r = await fetch(`${BASE}${p}`, {
    method,
    headers: { Authorization: `Bearer ${token()}`, "Content-Type": "application/json" },
    body: body ? JSON.stringify(body) : null,
  });
  const text = await r.text();
  let json; try { json = JSON.parse(text); } catch { return { status: r.status, body: { code: -1, raw: text } }; }
  return { status: r.status, body: json };
}

const payload = {
  element_name: name,
  element_description: description,
  reference_type: "image_refer",
  element_image_list: { frontal_image: frontal, refer_images: referList.map(u => ({ image_url: u })) },
  tag_list: [{ tag_id: tag }],
  external_task_id: externalId,
};

console.log(`▶ POST /v1/general/advanced-custom-elements  (name="${name}")`);
const submit = await http("POST", "/v1/general/advanced-custom-elements", payload);
if (submit.body.code !== 0) {
  console.error(`!! create failed: ${JSON.stringify(submit.body, null, 2)}`);
  process.exit(1);
}
const taskId = submit.body.data.task_id;
console.log(`>> task_id=${taskId}, polling...`);

const deadline = Date.now() + 5 * 60_000;
while (Date.now() < deadline) {
  await new Promise(r => setTimeout(r, 5_000));
  const poll = await http("GET", `/v1/general/advanced-custom-elements/${taskId}`);
  const d = poll.body.data || {};
  if (d.task_status === "succeed") {
    const el = ((d.task_result || {}).elements || [])[0];
    if (!el?.element_id) { console.error("succeed but no element"); process.exit(1); }
    const elementId = el.element_id;
    console.log(`\n✓ element_id=${elementId}  deduction=${d.final_unit_deduction}u`);

    // Append to registry
    const reg = fs.existsSync(REGISTRY_FILE) ? JSON.parse(fs.readFileSync(REGISTRY_FILE, "utf8")) : {};
    reg[name] = elementId;
    fs.writeFileSync(REGISTRY_FILE, JSON.stringify(reg, null, 2) + "\n");
    console.log(`✓ registry updated: ${path.relative(PROJECT_ROOT, REGISTRY_FILE)}`);
    process.exit(0);
  }
  if (d.task_status === "failed") {
    console.error(`!! create failed: ${JSON.stringify(d, null, 2)}`);
    process.exit(1);
  }
}
console.error("!! polling timeout"); process.exit(1);

File: saraandeva/.claude/skills/saraandeva-episode/scripts/pipeline/createElementViaApi.py
#!/usr/bin/env python3
"""
Create a Kling element via the official AK/SK API and append its
element_id to content/elements_registry.json.

Faithful Python port of createElementViaApi.mjs PLUS Fix B
(durability lesson_costume_element_coverage_gap.md): IDEMPOTENCY CHECK.
Before POSTing the create, list existing elements and search for matching
external_id. If found, skip the create and just write the existing
element_id to registry. Prevents the ep15-style duplicate problem where
the agent created 4 duplicate Kling elements because nobody checked first.

Usage:
  python3 createElementViaApi.py \\
    --name Mama \\
    --description "Adult woman, blonde hair, family-show character" \\
    --frontal https://storage.googleapis.com/saraandeva-kling-elements/characters/mama_front.png \\
    --refer https://.../mama_3q.png \\
    --refer https://.../mama_profile.png \\
    [--tag o_102]            # default Character
    [--external-id mama-1]
    [--force]                # ignore idempotency, always POST create

Tag IDs: o_101 Hottest, o_102 Character, o_103 Animal, o_104 Item,
         o_105 Costume, o_106 Scene, o_107 Effect, o_108 Others.

Note: Kling rejects refer_images with 0 entries — duplicate frontal if no
refers given.
"""
import argparse
import base64
import hashlib
import hmac
import json
import os
import sys
import time
from pathlib import Path
from urllib.request import Request, urlopen
from urllib.error import HTTPError

PROJECT_ROOT = Path("/Volumes/Samsung500/goreadling-production/saraandeva")
ENV_FILE = Path("/Volumes/Samsung500/goreadling-production/.env.local")
REGISTRY_FILE = PROJECT_ROOT / "content" / "elements_registry.json"
BASE = "https://api-singapore.klingai.com"
GENRE_CAP_NAME = 20
CAP_DESC = 100


def load_env():
    for line in ENV_FILE.read_text().splitlines():
        if "=" in line and not line.strip().startswith("#"):
            k, v = line.split("=", 1)
            os.environ.setdefault(k.strip(), v.strip().strip("'\""))
    ak, sk = os.environ.get("KLING_ACCESS_KEY"), os.environ.get("KLING_SECRET_KEY")
    if not ak or not sk:
        print("missing KLING keys in .env.local", file=sys.stderr); sys.exit(1)
    return ak, sk


def b64url(b: bytes) -> str:
    return base64.urlsafe_b64encode(b).rstrip(b"=").decode()


def make_token(ak: str, sk: str) -> str:
    now = int(time.time())
    h = b64url(json.dumps({"alg": "HS256", "typ": "JWT"}, separators=(",", ":")).encode())
    p = b64url(json.dumps({"iss": ak, "exp": now + 1800, "nbf": now - 5}, separators=(",", ":")).encode())
    s = b64url(hmac.new(sk.encode(), f"{h}.{p}".encode(), hashlib.sha256).digest())
    return f"{h}.{p}.{s}"


def http(method: str, path: str, ak: str, sk: str, body=None):
    headers = {"Authorization": f"Bearer {make_token(ak, sk)}", "Content-Type": "application/json"}
    data = json.dumps(body).encode() if body is not None else None
    req = Request(BASE + path, data=data, method=method, headers=headers)
    try:
        with urlopen(req, timeout=30) as r:
            text = r.read().decode()
            try: return r.status, json.loads(text)
            except json.JSONDecodeError: return r.status, {"code": -1, "raw": text[:200]}
    except HTTPError as e:
        try: return e.code, json.loads(e.read().decode())
        except Exception: return e.code, {"code": -1, "_http_error": str(e)}


# Fix B — idempotency search
def find_existing_by_external_id(ak: str, sk: str, external_id: str):
    """Search Kling library for existing element with matching external_id.
    Returns element_id or None."""
    for page in range(1, 11):
        status, body = http("GET",
            f"/v1/general/advanced-custom-elements?pageNum={page}&pageSize=500",
            ak, sk)
        if body.get("code") != 0: return None
        envelopes = body.get("data") or []
        if not envelopes: return None
        for env in envelopes:
            ti = env.get("task_info") or {}
            if ti.get("external_task_id") == external_id:
                el = ((env.get("task_result") or {}).get("elements") or [{}])[0]
                if el.get("element_id"):
                    return int(el["element_id"])
        if len(envelopes) < 500: return None
    return None


def write_registry(name: str, element_id: int):
    reg = json.loads(REGISTRY_FILE.read_text()) if REGISTRY_FILE.is_file() else {}
    reg[name] = element_id
    REGISTRY_FILE.write_text(json.dumps(reg, indent=2) + "\n")
    print(f"✓ registry updated: {REGISTRY_FILE.relative_to(PROJECT_ROOT)}")


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--name", required=True)
    ap.add_argument("--description", required=True)
    ap.add_argument("--frontal", required=True)
    ap.add_argument("--refer", action="append", default=[])
    ap.add_argument("--tag", default="o_102")
    ap.add_argument("--external-id", default=None)
    ap.add_argument("--force", action="store_true",
                    help="bypass idempotency check (always POST create)")
    args = ap.parse_args()

    if len(args.name) > GENRE_CAP_NAME:
        print(f"name > {GENRE_CAP_NAME} chars", file=sys.stderr); sys.exit(1)
    if len(args.description) > CAP_DESC:
        print(f"description > {CAP_DESC} chars", file=sys.stderr); sys.exit(1)

    refer_list = args.refer if args.refer else [args.frontal]
    if len(refer_list) > 3:
        print("max 3 refers", file=sys.stderr); sys.exit(1)

    external_id = args.external_id or f"{args.name.lower().replace(' ', '_')}-1"
    ak, sk = load_env()

    # ─── Fix B: idempotency check ────────────────────────────────────────
    if not args.force:
        print(f"▶ checking existing element by external_id={external_id}...")
        existing = find_existing_by_external_id(ak, sk, external_id)
        if existing:
            print(f"✓ existing element_id={existing} (external_id match) — skipping create")
            write_registry(args.name, existing)
            sys.exit(0)

    # ─── POST create ─────────────────────────────────────────────────────
    payload = {
        "element_name": args.name,
        "element_description": args.description,
        "reference_type": "image_refer",
        "element_image_list": {
            "frontal_image": args.frontal,
            "refer_images": [{"image_url": u} for u in refer_list],
        },
        "tag_list": [{"tag_id": args.tag}],
        "external_task_id": external_id,
    }
    print(f"▶ POST /v1/general/advanced-custom-elements  (name=\"{args.name}\")")
    status, body = http("POST", "/v1/general/advanced-custom-elements", ak, sk, payload)
    if body.get("code") != 0:
        print(f"!! create failed: {json.dumps(body, indent=2)[:600]}", file=sys.stderr); sys.exit(1)
    task_id = body["data"]["task_id"]
    print(f">> task_id={task_id}, polling...")

    deadline = time.time() + 5 * 60
    while time.time() < deadline:
        time.sleep(5)
        _, poll = http("GET", f"/v1/general/advanced-custom-elements/{task_id}", ak, sk)
        d = poll.get("data") or {}
        if d.get("task_status") == "succeed":
            els = (d.get("task_result") or {}).get("elements") or []
            if not els or not els[0].get("element_id"):
                print("succeed but no element", file=sys.stderr); sys.exit(1)
            eid = int(els[0]["element_id"])
            ded = d.get("final_unit_deduction", 0)
            print(f"\n✓ element_id={eid}  deduction={ded}u")
            write_registry(args.name, eid)
            sys.exit(0)
        if d.get("task_status") == "failed":
            print(f"!! create failed: {json.dumps(d, indent=2)[:600]}", file=sys.stderr); sys.exit(1)
    print("!! polling timeout", file=sys.stderr); sys.exit(1)


if __name__ == "__main__":
    main()
